lesson04.md

```diff
@@ -0,0 +1,460 @@
+<!-- .slide: id="lesson4" -->
+
+# Basic Frontend - Fall 2021
+
+Lesson 4, Thursday, 2021-09-23
+
+---
+
+### Lesson overview
+
+* `undefined` datatype
+* Operator precedence
+* Operator chaining
+* Conditional statements
+
+---
+
+### Recap
+
+Name the operators:
+
+```js
+&&
+<
+===
+||
++
+**
+```
+
+---
+
+### Recap
+
+Name the operators:
+
+```js
+&&    // logical AND
+<     // less than
+===   // strict equality
+||    // logical OR
++     // addition
+**    // exponentiation
+```
+
+---
+
+### Operators and data types
+
+| Data Type | Operators | Example |
+| ----- | --------- | ------- |
+| `number` | `+ - * / **`<br/>`=== !== < > <= >=` | `5 + 4 * 3`<br/>`7 / 2 - 2`<br/>`32 > 45` |
+| `string` | `+ === !==` | `"Pi" === "Pa"`<br/>`"Hello" + " World"` |
+| `boolean` | <code>&#124;&#124; && ! === !==</code> | `true && !false` |
+
+---
+
+### What does it print?
+
+```js
+let name = "Sevtap";
+console.log("Hello " + name);
+```
+
+Answer: It prints "Hello Sevtap" <!-- .element: class="fragment" -->
+
+---
+
+### What does it print?
+
+```js
+let greeting = "Hello";
+let name = "Sevtap";
+console.log(greeting + " " + firstName);
+```
+
+Answer: Error, "firstName" is not defined <!-- .element: class="fragment" -->
+
+---
+
+### What does it print?
+
+```js
+console.log("It is " + temperature + " degrees");
+let temperature = 22;
+```
+
+Answer: Error, "temperature" is not defined <!-- .element: class="fragment" -->
+
+---
+
+### Data Type: Undefined
+
+* **Undefined** has only one value: `undefined`
+* It indicates the unintentional absence of any value
+* `undefined` is automatically assigned to variables
+
+---
+
+### Undefined example
+
+```js
+let x;  // no value is assigned to x
+console.log(x); // undefined
+console.log(typeof x); // "undefined"
+console.log(x === undefined); // true
+````
+
+---
+
+### Operator precedence
+
+What do you think is the result?
+
+```js
+2 + 2 * 2
+```
+
+Result: `6` <!-- .element: class="fragment" -->
+
+---
+
+### Operator precedence (cont.)
+
+* JavaScript supports precedence (priority) for operators
+* Operators with higher precedence are evaluated before operators with lower precedence
+* Full List: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence
+* Multiplication (15) has higher precedence than addition (14) - multiplication wins
+
+---
+
+### Operator precedence (cont.)
+
+* Parentheses have the highest precedence (21) and always win:
+
+```js
+(2 + 2) * 2   // 8
+2 + (2 * 2)   // 6
+```
+
+* When in doubt, use parentheses!
+
+---
+
+### Quiz
+
+What is `x` ?
+
+```js
+let x = 4 + 6 < 3 + 8;
+```
+
+Result: `true` <!-- .element: class="fragment" -->
+
+---
+
+### Best practice: Use extra variables
+
+Which of the two is more readable?
+
+```js
+let x = 4 + 6 < 3 + 8;
+```
+
+```js
+let sum1 = 4 + 6;
+let sum2 = 3 + 6;
+let x = sum1 < sum2;
+```
+
+---
+
+### Operator chaining
+
+We can "chain" operators, for example:
```

Comment:
What happens when two operators have the same precedence value, which one gets evaluated first. In this case we look at the associativity rule of the operator. If an operator has left-to-right associativity, then we evaluate the expression on the left of the common value first. Below 2 is the common value, and left associativity tells us to compute 5 + 2 first.

Maybe a picture will make it clearer. `5-2+3`
https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcTu09t6a9zJXO1jQPNGcTK0CyX3wKl06O-sRw&usqp=CAU

Reply:
* Wost of the operators are left-to-right associative.

* Assignment operator has right-to-left associativity, which means we evaluate things on the right side of the `=` operator first before we change the value on the left side of the operator.

* When in doubt consult the MDN documentation.